repo_policy_compliance/github_client.py

```diff
@@ -53,7 +67,119 @@ def get() -> Github:
         raise_on_status=False,
         raise_on_redirect=False,
     )
-    return Github(auth=Token(github_token), retry=retry_config)
+    return Github(auth=auth, retry=retry_config)
+
+
+def _get_auth() -> Auth:
+    """Get a GitHub auth object.
+
+    Returns:
+        A GitHub auth object that is configured with a token from the environment.
+    """
+    github_token = os.getenv(GITHUB_TOKEN_ENV_NAME) or os.getenv(f"FLASK_{GITHUB_TOKEN_ENV_NAME}")
+    github_app_id = os.getenv(GITHUB_APP_ID_ENV_NAME) or os.getenv(
+        f"FLASK_{GITHUB_APP_ID_ENV_NAME}"
+    )
+    github_app_installation_id_str = os.getenv(GITHUB_APP_INSTALLATION_ID_ENV_NAME) or os.getenv(
+        f"FLASK_{GITHUB_APP_INSTALLATION_ID_ENV_NAME}"
+    )
+    github_app_private_key = os.getenv(GITHUB_APP_PRIVATE_KEY_ENV_NAME) or os.getenv(
+        f"FLASK_{GITHUB_APP_PRIVATE_KEY_ENV_NAME}"
+    )
+
+    _ensure_either_github_token_or_app_config(
+        github_token=github_token,
+        github_app_id=github_app_id,
+        github_app_installation_id_str=github_app_installation_id_str,
+        github_app_private_key=github_app_private_key,
+    )
```

Comment:
You can restructure this a bit so that you don't need to have exceptions that could bubble up and asserts etc. You can have a function like this which determines which authentication method can be used which has 3 cases (can be returned as an enum):

* token based auth
* app based auth
* auth error (could be a few different errors to indicate what is wrong, e.g., no auth provided, some of the app auth variables are missing, ...)

Then after this you can do a switch on the enum and use the environ dict and assume the required variables are there

Reply:
I think we've had discussion on this point before (@gregory-schiano ) about when to use ValidationResult vs raise Exception.
My opinion is that since we're not using Validation result, we should throw an error.
If we were to return enums (I think this is a good idea), the function should change (perhaps `_get_auth_mode`) and the return values should be two token/app based enum but the exception should still be thrown.
What do you think?

Reply:
I agree with Yanks here. I have updated the code to return an enum and removed the asserts (they were/are necessary for mypy, but can be replaced with casts). The method still throws an error on invalid configuration.